airflow/models/dag.py

```diff
@@ -2648,323 +2643,38 @@ def bulk_write_to_db(
-            )
-            if stored_dataset_model:
-                # Some datasets may have been previously unreferenced, and therefore orphaned by the
-                # scheduler. But if we're here, then we have found that dataset again in our DAGs, which
-                # means that it is no longer an orphan, so set is_orphaned to False.
-                stored_dataset_model.is_orphaned = expression.false()
-                stored_dataset_models[stored_dataset_model.uri] = stored_dataset_model
-            else:
-                new_dataset_models.append(dataset)
-        dataset_manager.create_datasets(dataset_models=new_dataset_models, session=session)
-        stored_dataset_models.update(
-            {dataset_model.uri: dataset_model for dataset_model in new_dataset_models}
-        )
-
-        del new_dataset_models
-        del all_dataset_models
-
-        # store dataset aliases
-        all_datasets_alias_models = input_dataset_alias_models | outlet_dataset_alias_models
-        stored_dataset_alias_models: dict[str, DatasetAliasModel] = {}
-        new_dataset_alias_models: set[DatasetAliasModel] = set()
-        if all_datasets_alias_models:
-            all_dataset_alias_names = {
-                dataset_alias_model.name for dataset_alias_model in all_datasets_alias_models
-            }
-
-            stored_dataset_alias_models = {
-                dsa_m.name: dsa_m
-                for dsa_m in session.scalars(
-                    select(DatasetAliasModel).where(DatasetAliasModel.name.in_(all_dataset_alias_names))
-                ).fetchall()
-            }
-
-            if stored_dataset_alias_models:
-                new_dataset_alias_models = {
-                    dataset_alias_model
-                    for dataset_alias_model in all_datasets_alias_models
-                    if dataset_alias_model.name not in stored_dataset_alias_models.keys()
-                }
-            else:
-                new_dataset_alias_models = all_datasets_alias_models
-
-            session.add_all(new_dataset_alias_models)
-        session.flush()
-        stored_dataset_alias_models.update(
-            {
-                dataset_alias_model.name: dataset_alias_model
-                for dataset_alias_model in new_dataset_alias_models
-            }
+        log.info("Sync %s DAGs", len(dags))
+        dags_by_ids = {dag.dag_id: dag for dag in dags}
+        del dags
+
+        orm_dags = collect_orm_dags(dags_by_ids, session=session)
+        orm_dags.update(
+            (dag_id, create_orm_dag(dag, session=session))
+            for dag_id, dag in dags_by_ids.items()
+            if dag_id not in orm_dags
         )
 
-        del new_dataset_alias_models
-        del all_datasets_alias_models
+        update_orm_dags(dags_by_ids, orm_dags, processor_subdir=processor_subdir, session=session)
+        DagCode.bulk_sync_to_db((dag.fileloc for dag in dags_by_ids.values()), session=session)
```

Comment:
Should we somehow return the fileloc instead of iterating through the dags again to get the filelocs?

Reply:
Performance should be roughly the same since `dags_by_ids` is in-memory. I do plan to further refactor this in a later PR, as mentioned in https://github.com/apache/airflow/pull/42245#discussion_r1766097408. This will be done in a similar fashion as DatasetCollection.